### PR TITLE
[HotFix] Fix stage_index in zerobubble test;

### LINKED ISSUE #6037: [FEATURE]: Support Zerobubble pipeline
### Describe the feature

This issue contains requirements for the zerobubble pp.
Source Github Link: https://github.com/sail-sg/zero-bubble-pipeline-parallelism
Source Paper: https://arxiv.org/abs/2401.10241

### PR DESCRIPTION
**Pull request by duanjunwen**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs
- [x] I have installed pre-commit: `pip install pre-commit && pre-commit install`

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

resolved #6037

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/pipeline/schedule/zero_bubble_pp.py [hunk 1]
```diff
@@ -480,22 +480,26 @@ def backward_b_step(
...
         # For chunk 0 stage 0, use micro_batch as input_obj_
         if model_chunk_id == 0 and self.stage_manager.is_first_stage(ignore_chunk=True):
-            input_obj_, _ = tree_flatten(micro_batch)
-            output_obj_, _ = tree_flatten(output_obj)  # y
-            output_obj_grad_, _ = tree_flatten(output_obj_grad)  # dy
+            input_obj_, _ = tree_flatten({k: v for k, v in micro_batch.items() if isinstance(v, torch.Tensor)})
+            output_obj_, _ = tree_flatten({k: v for k, v in output_obj.items() if isinstance(v, torch.Tensor)})  # y
+            output_obj_grad_, _ = tree_flatten(
+                {k: v for k, v in output_obj_grad.items() if isinstance(v, torch.Tensor)}
+            )  # dy
```
Comment (ver217) [inline colossalai/pipeline/schedule/zero_bubble_pp.py:487]: don't traverse such obj like dict, just use tree flatten and traverse the list

Comment (duanjunwen) [reply colossalai/pipeline/schedule/zero_bubble_pp.py]: > don't traverse such obj like dict, just use tree flatten and traverse the list

Fixed in https://github.com/hpcaitech/ColossalAI/pull/6069/commits/83163fa70c49085b15ea063fcb0ee188d28f4871；

File: tests/test_pipeline/test_schedule/test_zerobubble_pp.py [hunk 2]
```diff
@@ -22,36 +21,51 @@
...
+        stage_mgr: PipelineStageManager = None,
+    ):
         super().__init__()
-        self.layers = nn.ModuleList([nn.Linear(in_dim, out_dim, bias=None) for _ in range(num_layers)])
+        self.layers = nn.Sequential(*[nn.Linear(in_dim, out_dim, bias=None) for _ in range(num_layers)])
 
     def forward(
         self,
-        hidden_states,
+        model=None,
```
[the remark this answers is not in the document]
Comment (duanjunwen) [reply tests/test_pipeline/test_schedule/test_zerobubble_pp.py]: Removed in https://github.com/hpcaitech/ColossalAI/pull/6069/commits/bb0390c90d8645b2d58035e82335049c468d36ec;

[the remark this answers is not in the document]
Comment (duanjunwen) [reply tests/test_pipeline/test_schedule/test_zerobubble_pp.py]: > can we remove this now?

Removed in https://github.com/hpcaitech/ColossalAI/pull/6069/commits/bb0390c90d8645b2d58035e82335049c468d36ec;

**Review by ver217 — APPROVED** [review]

**Pull request merged**